backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+		FROM durabletask.Instances WHERE InstanceID = $1`,
+		string(iid),
+	)
+
+	err := row.Err()
+	if err == sql.ErrNoRows {
+		return nil, api.ErrInstanceNotFound
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to query the Instances table: %w", row.Err())
+	}
+
+	var instanceID *string
+	var name *string
+	var runtimeStatus *string
+	var createdAt *time.Time
+	var lastUpdatedAt *time.Time
+	var input *string
+	var output *string
+	var customStatus *string
+	var failureDetails *protos.TaskFailureDetails
+
+	var failureDetailsPayload []byte
+	err = row.Scan(&instanceID, &name, &runtimeStatus, &createdAt, &lastUpdatedAt, &input, &output, &customStatus, &failureDetailsPayload)
+	if err == sql.ErrNoRows {
+		return nil, api.ErrInstanceNotFound
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to scan the Instances table result: %w", err)
+	}
+
+	if input == nil {
+		input = &emptyString
+	}
+
+	if output == nil {
+		output = &emptyString
+	}
+
+	if customStatus == nil {
+		customStatus = &emptyString
+	}
+
+	if len(failureDetailsPayload) > 0 {
+		failureDetails = new(protos.TaskFailureDetails)
+		if err := protojson.Unmarshal(failureDetailsPayload, failureDetails); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal failure details: %w", err)
+		}
+	}
+
+	metadata := api.NewOrchestrationMetadata(
+		iid,
+		*name,
+		helpers.FromRuntimeStatusString(*runtimeStatus),
+		*createdAt,
+		*lastUpdatedAt,
+		*input,
+		*output,
+		*customStatus,
+		failureDetails,
+	)
+	return metadata, nil
+}
+
+// GetOrchestrationRuntimeState implements backend.Backend
+func (be *postgresBackend) GetOrchestrationRuntimeState(ctx context.Context, wi *backend.OrchestrationWorkItem) (*backend.OrchestrationRuntimeState, error) {
+	if err := be.ensureDB(ctx); err != nil {
+		return nil, err
+	}
+
+	rows, err := be.db.QueryContext(
+		ctx,
+		"SELECT EventPayload FROM durabletask.History WHERE InstanceID = $1 ORDER BY SequenceNumber ASC",
+		string(wi.InstanceID),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	existingEvents := make([]*protos.HistoryEvent, 0, 50)
+	for rows.Next() {
+		var eventPayload []byte
+		if err := rows.Scan(&eventPayload); err != nil {
+			return nil, fmt.Errorf("failed to read history event: %w", err)
+		}
+		e := new(protos.HistoryEvent)
+		err := protojson.Unmarshal(eventPayload, e)
+		if err != nil {
+			return nil, err
+		}
+
+		existingEvents = append(existingEvents, e)
+	}
+
+	state := backend.NewOrchestrationRuntimeState(wi.InstanceID, existingEvents)
+	return state, nil
+}
+
+// GetOrchestrationWorkItem implements backend.Backend
+func (be *postgresBackend) GetOrchestrationWorkItem(ctx context.Context) (*backend.OrchestrationWorkItem, error) {
+	if err := be.ensureDB(ctx); err != nil {
+		return nil, err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	now := time.Now().UTC()
+	newLockExpiration := now.Add(be.orchestrationLockTimeout)
```

Comment:
For all operations with times, there's always the potential that issues may arise due to clock skews between multiple nodes.

A pattern I would recommend is to pass the delay (`be.orchestrationLockTimeout`) to Postgres, and let Postgres do the math with the current time (`E.VisibleTime < CURRENT_TIMESTAMP + $1::interval`)